distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso

```diff
@@ -0,0 +1,158 @@
+        if this.name.is_nothing.not then this.name else
+            col_name = c->(this.resolve_column table c).name
+            case this of
+                Count _ -> "Count"
+                Count_Distinct columns _ -> "Count Distinct " + (columns.map col_name . join " ")
+                Count_Not_Null c _ -> "Count NotNull " + col_name c
+                Count_Null c _ -> "Count Null " + col_name c
+                Count_Not_Empty c _ -> "Count NotEmpty " + col_name c
+                Count_Empty c _ -> "Count Empty " + col_name c
+                Sum c _ -> "Sum " + col_name c
+                Average c _ -> "Average " + col_name c
+                Median c _ -> "Median " + col_name c
+                Standard_Deviation c _ _ -> "Standard Deviation " + col_name c
+                Concatenate c _ _ _ _ -> "Concatenate " + col_name c
+                First c _ -> "First " + col_name c
+                Last c _ -> "Last " + col_name c
+
+    ## PRIVATE
+       Given a column reference resolve to the underlying column
+    resolve_column : Table->(Column|Text|Integer)->Column
+    resolve_column table column = case column of
+        Column _ -> table.at (column.name)
+        Text -> table.at column
+        Integer -> table.columns.at column
+
+    initial_value : Any
+    initial_value = case this of
+        Count_Distinct _ _ -> Map.empty
+        Median _ _ -> Map.empty
+        Average _ _ -> [0, 0]
+        Standard_Deviation _ _ _ -> [0, 0, 0]
+        Concatenate _ _ _ _ _ -> Nothing
+        First _ _ -> Nothing
+        Last _ _ -> Nothing
+        _ -> 0
+
+    make_aggregator : Table->(Any->Integer->Any)
+    make_aggregator table =
+        create_closure c function =
+            col = this.resolve_column table c
+            function col
+
+        is_empty s = if s.is_nothing then True else case s of
+            Text -> s.is_empty
+            _ -> Error.throw (Invalid_Aggregation_Method this.col "Empty is only valid for Text")
+
+        case this of
+            Count _ -> count->_->(count+1)
+            Count_Not_Null c _ -> create_closure c col->count->i->(count + if (col.at i).is_nothing then 0 else 1)
+            Count_Null c _ -> create_closure c col->count->i->(count + if (col.at i).is_nothing then 1 else 0)
+            Count_Not_Empty c _ -> create_closure c col->count->i->(count + if is_empty (col.at i) then 0 else 1)
+            Count_Empty c _ -> create_closure c col->count->i->(count + if is_empty (col.at i) then 1 else 0)
+            First c _ -> create_closure c col->current->i->(if i==0 then (col.at i) else current)
+            Last c _ -> create_closure c col->_->i->(col.at i)
+            Sum c _ -> create_closure c col->total->i->
+                v = col.at i
+                (total + if v.is_nothing then 0 else v)
+            Average c _ -> create_closure c col->a->i->
+                v = col.at i
+                (c + if v.is_nothing then a else [(a.at 0) + 1, (a.at 1) + v])
```

Comment:
```suggestion
                (c + if v.is_nothing then a else [a.first + 1, a.second + v])
```

I think it may be worth using these two helper functions, will be clearer to read (at least in my opinion - because when I see `at` I usually think it may be an arbitrary size array and if I see `first` and `second` I immediately connect this with the accumulator which is a pair).